go/vt/vttablet/tabletserver/vstreamer/packet_size.go

```diff
@@ -0,0 +1,184 @@
+	return &fixedPacketSizer{baseSize: baseSize}
+}
+
+func (ps *fixedPacketSizer) Limit() int {
+	return ps.baseSize
+}
+
+// ShouldSend checks whether the given byte count is large enough to be sent as a packet while streaming
+func (ps *fixedPacketSizer) ShouldSend(byteCount int) bool {
+	return byteCount >= ps.baseSize
+}
+
+// Record records the total duration it took to send the given byte count while streaming
+func (ps *fixedPacketSizer) Record(_ int, _ time.Duration) {}
+
+type dynamicPacketSizer struct {
+	last      int
+	grow      int
+	target    int
+	hits      int
+	settled   bool
+	current   *mathstats.Sample
+	candidate *mathstats.Sample
+	elapsed   time.Duration
+}
+
+func newDynamicPacketSizer(baseSize int) PacketSizer {
+	return &dynamicPacketSizer{
+		last:      baseSize,
+		current:   &mathstats.Sample{},
+		candidate: &mathstats.Sample{},
+		target:    baseSize,
+		grow:      baseSize / 4,
+	}
+}
+
+func (ps *dynamicPacketSizer) Limit() int {
+	return ps.target
+}
+
+// ShouldSend checks whether the given byte count is large enough to be sent as a packet while streaming
+func (ps *dynamicPacketSizer) ShouldSend(byteCount int) bool {
+	return byteCount >= ps.target
+}
+
+type change int8
+
+const (
+	notChanging change = iota
+	gettingFaster
+	gettingSlower
+)
+
+func (ps *dynamicPacketSizer) changeInThroughput() change {
+	const PValueMargin = 0.1
+
+	t, err := mathstats.TwoSampleWelchTTest(ps.current, ps.candidate, mathstats.LocationDiffers)
+	if err != nil {
+		return notChanging
+	}
+	if t.P < PValueMargin {
+		if ps.candidate.Mean() > ps.current.Mean() {
+			return gettingFaster
+		}
+		return gettingSlower
+	}
+	return notChanging
+}
+
+func (ps *dynamicPacketSizer) reset() {
+	ps.current.Clear()
+	ps.candidate.Clear()
+	ps.hits = 0
+	ps.settled = false
+	ps.elapsed = 0
+}
+
+// Record records the total duration it took to send the given byte count while streaming
+func (ps *dynamicPacketSizer) Record(byteCount int, d time.Duration) {
+	const ExperimentDelay = 5 * time.Second
+	const CheckFrequency = 16
+	const GrowthFrequency = 32
+	const InitialCandidateLen = 32
+	const SettleCandidateLen = 64
+
+	if ps.settled {
+		ps.elapsed += d
+		if ps.elapsed < ExperimentDelay {
+			return
+		}
+		ps.reset()
+	}
+
+	ps.hits++
+	ps.candidate.Xs = append(ps.candidate.Xs, float64(byteCount)/float64(d))
+	if ps.hits%CheckFrequency == 0 {
+		ps.candidate.Sorted = false
+		ps.candidate.FilterOutliers()
+
+		if len(ps.current.Xs) == 0 {
+			if len(ps.candidate.Xs) >= InitialCandidateLen {
+				ps.current, ps.candidate = ps.candidate, ps.current
+				log.Infof("packetSize: stored initial sample, mean = %f", ps.current.Mean())
+			}
+			return
+		}
+
+		change := ps.changeInThroughput()
+		switch change {
+		case notChanging, gettingSlower:
+			if len(ps.candidate.Xs) >= SettleCandidateLen {
+				ps.target = ps.last
+				ps.settled = true
+				log.Infof("packetSize: not changing for %d, settled at %d", len(ps.candidate.Xs), ps.target)
```

Comment:
How often will these statements log? If it is going to happen a lot we should probably comment them out since they can spam the logs.